// @azure/monitor-opentelemetry/dist-esm/src/types.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
export const AZURE_MONITOR_STATSBEAT_FEATURES = "AZURE_MONITOR_STATSBEAT_FEATURES";
export var StatsbeatFeature;
(function (StatsbeatFeature) {
    StatsbeatFeature[StatsbeatFeature["DISK_RETRY"] = 0] = "DISK_RETRY";
    StatsbeatFeature[StatsbeatFeature["AAD_HANDLING"] = 1] = "AAD_HANDLING";
    StatsbeatFeature[StatsbeatFeature["WEB_SNIPPET"] = 2] = "WEB_SNIPPET";
    StatsbeatFeature[StatsbeatFeature["DISTRO"] = 4] = "DISTRO";
})(StatsbeatFeature || (StatsbeatFeature = {}));
export var StatsbeatInstrumentation;
(function (StatsbeatInstrumentation) {
    StatsbeatInstrumentation[StatsbeatInstrumentation["AZURE_CORE_TRACING"] = 0] = "AZURE_CORE_TRACING";
    StatsbeatInstrumentation[StatsbeatInstrumentation["MONGODB"] = 1] = "MONGODB";
    StatsbeatInstrumentation[StatsbeatInstrumentation["MYSQL"] = 2] = "MYSQL";
    StatsbeatInstrumentation[StatsbeatInstrumentation["REDIS"] = 4] = "REDIS";
    StatsbeatInstrumentation[StatsbeatInstrumentation["POSTGRES"] = 8] = "POSTGRES";
})(StatsbeatInstrumentation || (StatsbeatInstrumentation = {}));
