// @azure/monitor-opentelemetry/dist-esm/src/metrics/types.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
export var StandardMetricNames;
(function (StandardMetricNames) {
    StandardMetricNames["HTTP_REQUEST_DURATION"] = "azureMonitor.http.requestDuration";
    StandardMetricNames["HTTP_DEPENDENCY_DURATION"] = "azureMonitor.http.dependencyDuration";
    StandardMetricNames["EXCEPTION_COUNT"] = "azureMonitor.exceptionCount";
    StandardMetricNames["TRACE_COUNT"] = "azureMonitor.traceCount";
})(StandardMetricNames || (StandardMetricNames = {}));
export var PerformanceCounterMetricNames;
(function (PerformanceCounterMetricNames) {
    PerformanceCounterMetricNames["PRIVATE_BYTES"] = "\\Process(??APP_WIN32_PROC??)\\Private Bytes";
    PerformanceCounterMetricNames["AVAILABLE_BYTES"] = "\\Memory\\Available Bytes";
    PerformanceCounterMetricNames["PROCESSOR_TIME"] = "\\Processor(_Total)\\% Processor Time";
    PerformanceCounterMetricNames["PROCESS_TIME"] = "\\Process(??APP_WIN32_PROC??)\\% Processor Time";
    PerformanceCounterMetricNames["REQUEST_RATE"] = "\\ASP.NET Applications(??APP_W3SVC_PROC??)\\Requests/Sec";
    PerformanceCounterMetricNames["REQUEST_DURATION"] = "\\ASP.NET Applications(??APP_W3SVC_PROC??)\\Request Execution Time";
})(PerformanceCounterMetricNames || (PerformanceCounterMetricNames = {}));
