// applicationinsights/out/src/shim/logging/logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Logger = void 0;
const api_1 = require("@opentelemetry/api");
const internalAzureLogger_1 = require("./internalAzureLogger");
class Logger {
    constructor() {
        this._TAG = "ApplicationInsights:";
        this._internalLogger = new internalAzureLogger_1.InternalAzureLogger();
        const envLogLevel = process.env.APPLICATIONINSIGHTS_INSTRUMENTATION_LOGGING_LEVEL;
        this._diagLevel = api_1.DiagLogLevel.WARN; // Default
        switch (envLogLevel) {
            case "ALL":
                this._diagLevel = api_1.DiagLogLevel.ALL;
                break;
            case "DEBUG":
                this._diagLevel = api_1.DiagLogLevel.DEBUG;
                break;
            case "ERROR":
                this._diagLevel = api_1.DiagLogLevel.ERROR;
                break;
            case "INFO":
                this._diagLevel = api_1.DiagLogLevel.INFO;
                break;
            case "NONE":
                this._diagLevel = api_1.DiagLogLevel.NONE;
                break;
            case "VERBOSE":
                this._diagLevel = api_1.DiagLogLevel.VERBOSE;
                break;
            case "WARN":
                this._diagLevel = api_1.DiagLogLevel.WARN;
                break;
        }
        this.updateLogLevel(this._diagLevel);
    }
    static getInstance() {
        if (!Logger._instance) {
            Logger._instance = new Logger();
        }
        return Logger._instance;
    }
    /**
     * Set the global LogLevel. If a global diag logger is already set, this will override it.
     * @param logLevel - The DiagLogLevel used to filter logs sent to the logger.
     * @param suppressOverrideMessage - Setting that suppress the warning message normally emitted when registering a logger when another logger is already registered.
     */
    updateLogLevel(logLevel, suppressOverrideMessage = true) {
        this._diagLevel = logLevel;
        // Set OpenTelemetry Logger
        api_1.diag.setLogger(this, {
            logLevel: this._diagLevel,
            suppressOverrideMessage,
        });
    }
    error(message, ...optionalParams) {
        if (this._diagLevel >= api_1.DiagLogLevel.ERROR) {
            this._internalLogger.logMessage(this._TAG + message, optionalParams);
        }
    }
    warn(message, ...optionalParams) {
        if (this._diagLevel >= api_1.DiagLogLevel.WARN) {
            this._internalLogger.logMessage(this._TAG + message, optionalParams);
        }
    }
    info(message, ...optionalParams) {
        if (this._diagLevel >= api_1.DiagLogLevel.INFO) {
            this._internalLogger.logMessage(this._TAG + message, optionalParams);
        }
    }
    debug(message, ...optionalParams) {
        if (this._diagLevel >= api_1.DiagLogLevel.DEBUG) {
            this._internalLogger.logMessage(this._TAG + message, optionalParams);
        }
    }
    verbose(message, ...optionalParams) {
        if (this._diagLevel >= api_1.DiagLogLevel.VERBOSE) {
            this._internalLogger.logMessage(this._TAG + message, optionalParams);
        }
    }
}
exports.Logger = Logger;
